### ISSUE TITLE
使用 internlm2-chat-1.8b模型微调后，大概率出现了 循环输出的情况。

### ISSUE THREAD

**Issue by zhanghui-china** [InternLM/xtuner#446]
训练配置：
···
# Copyright (c) OpenMMLab. All rights reserved.
import torch
from datasets import load_dataset
from mmengine.dataset import DefaultSampler
from mmengine.hooks import (CheckpointHook, DistSamplerSeedHook, IterTimerHook,
                            LoggerHook, ParamSchedulerHook)
from mmengine.optim import AmpOptimWrapper, CosineAnnealingLR, LinearLR
from peft import LoraConfig
from torch.optim import AdamW
from transformers import (AutoModelForCausalLM, AutoTokenizer,
                          BitsAndBytesConfig)

from xtuner.dataset import process_hf_dataset
from xtuner.dataset.collate_fns import default_collate_fn
from xtuner.dataset.map_fns import oasst1_map_fn, template_map_fn_factory
from xtuner.engine.hooks import (DatasetInfoHook, EvaluateChatHook,
                                 VarlenAttnArgsToMessageHubHook)
from xtuner.engine.runner import TrainLoop
from xtuner.model import SupervisedFinetune
from xtuner.utils import PROMPT_TEMPLATE

#######################################################################
#                          PART 1  Settings                           #
#######################################################################
# Model
pretrained_model_name_or_path = '/home/zhanghui/models/Shanghai_AI_Laboratory/internlm2-chat-1_8b'
use_varlen_attn = False

# Data
data_path = './dataset/tran_dataset_0.json'
prompt_template = PROMPT_TEMPLATE.internlm2_chat
max_length = 2048
pack_to_max_length = True

# Scheduler & Optimizer
batch_size = 1  # per_device
accumulative_counts = 16
dataloader_num_workers = 0
max_epochs = 1
optim_type = AdamW
lr = 2e-4
betas = (0.9, 0.999)
weight_decay = 0
max_norm = 1  # grad clip
warmup_ratio = 0.03

# Save
save_steps = 500
save_total_limit = 2  # Maximum checkpoints to keep (-1 means unlimited)

# Evaluate the generation performance during the training
evaluation_freq = 500
SYSTEM = ''
evaluation_inputs = [
    '请给我介绍五个上海的景点', 'Please tell me five scenic spots in Shanghai'
]

#######################################################################
#                      PART 2  Model & Tokenizer                      #
#######################################################################
tokenizer = dict(
    type=AutoTokenizer.from_pretrained,
    pretrained_model_name_or_path=pretrained_model_name_or_path,
    trust_remote_code=True,
    padding_side='right')

model = dict(
    type=SupervisedFinetune,
    use_varlen_attn=use_varlen_attn,
    llm=dict(
        type=AutoModelForCausalLM.from_pretrained,
        pretrained_model_name_or_path=pretrained_model_name_or_path,
        trust_remote_code=True,
        torch_dtype=torch.float16,
        quantization_config=dict(
            type=BitsAndBytesConfig,
            load_in_4bit=True,
            load_in_8bit=False,
            llm_int8_threshold=6.0,
            llm_int8_has_fp16_weight=False,
            bnb_4bit_compute_dtype=torch.float16,
            bnb_4bit_use_double_quant=True,
            bnb_4bit_quant_type='nf4')),
    lora=dict(
        type=LoraConfig,
        r=64,
        lora_alpha=16,
        lora_dropout=0.1,
        bias='none',
        task_type='CAUSAL_LM'))

#######################################################################
#                      PART 3  Dataset & Dataloader                   #
#######################################################################
train_dataset = dict(
    type=process_hf_dataset,
    #dataset=dict(type=load_dataset, path=data_path),
    dataset=dict(type=load_dataset, path='json', data_files=dict(train=data_path)),
    tokenizer=tokenizer,
    max_length=max_length,
    #dataset_map_fn=oasst1_map_fn,
    dataset_map_fn=None,
    template_map_fn=dict(
        type=template_map_fn_factory, template=prompt_template),
    remove_unused_columns=True,
    shuffle_before_pack=True,
    pack_to_max_length=pack_to_max_length,
    use_varlen_attn=use_varlen_attn)

train_dataloader = dict(
    batch_size=batch_size,
    num_workers=dataloader_num_workers,
    dataset=train_dataset,
    sampler=dict(type=DefaultSampler, shuffle=True),
    collate_fn=dict(type=default_collate_fn, use_varlen_attn=use_varlen_attn))

#######################################################################
#                    PART 4  Scheduler & Optimizer                    #
#######################################################################
# optimizer
optim_wrapper = dict(
    type=AmpOptimWrapper,
    optimizer=dict(
        type=optim_type, lr=lr, betas=betas, weight_decay=weight_decay),
    clip_grad=dict(max_norm=max_norm, error_if_nonfinite=False),
    accumulative_counts=accumulative_counts,
    loss_scale='dynamic',
    dtype='float16')

# learning policy
# More information: https://github.com/open-mmlab/mmengine/blob/main/docs/en/tutorials/param_scheduler.md  # noqa: E501
param_scheduler = [
    dict(
        type=LinearLR,
        start_factor=1e-5,
        by_epoch=True,
        begin=0,
        end=warmup_ratio * max_epochs,
        convert_to_iter_based=True),
    dict(
        type=CosineAnnealingLR,
        eta_min=0.0,
        by_epoch=True,
        begin=warmup_ratio * max_epochs,
        end=max_epochs,
        convert_to_iter_based=True)
]

# train, val, test setting
train_cfg = dict(type=TrainLoop, max_epochs=max_epochs)

#######################################################################
#                           PART 5  Runtime                           #
#######################################################################
# Log the dialogue periodically during the training process, optional
custom_hooks = [
    dict(type=DatasetInfoHook, tokenizer=tokenizer),
    dict(
        type=EvaluateChatHook,
        tokenizer=tokenizer,
        every_n_iters=evaluation_freq,
        evaluation_inputs=evaluation_inputs,
        system=SYSTEM,
        prompt_template=prompt_template)
]

if use_varlen_attn:
    custom_hooks += [dict(type=VarlenAttnArgsToMessageHubHook)]

# configure default hooks
default_hooks = dict(
    # record the time of every iteration.
    timer=dict(type=IterTimerHook),
    # print log every 10 iterations.
    logger=dict(type=LoggerHook, log_metric_by_epoch=False, interval=10),
    # enable the parameter scheduler.
    param_scheduler=dict(type=ParamSchedulerHook),
    # save checkpoint per `save_steps`.
    checkpoint=dict(
        type=CheckpointHook,
        by_epoch=False,
        interval=save_steps,
        max_keep_ckpts=save_total_limit),
    # set sampler seed in distributed evrionment.
    sampler_seed=dict(type=DistSamplerSeedHook),
)

# configure environment
env_cfg = dict(
    # whether to enable cudnn benchmark
    cudnn_benchmark=False,
    # set multi process parameters
    mp_cfg=dict(mp_start_method='fork', opencv_num_threads=0),
    # set distributed parameters
    dist_cfg=dict(backend='nccl'),
)

# set visualizer
visualizer = None

# set log level
log_level = 'INFO'

# load from which checkpoint
load_from = None

# whether to resume training from the loaded checkpoint
resume = False

# Defaults to use random seed and disable `deterministic`
randomness = dict(seed=None, deterministic=False)

# set log processor
log_processor = dict(by_epoch=False)
···
训练：
xtuner train ./internlm2_chat_1_8b_qlora_copy.py --deepspeed deepspeed_zero2
![image](https://github.com/InternLM/xtuner/assets/63148804/069cef22-f93b-4a79-88ee-a050569a84cc)

生成Lora模型：
rm -rf hf

mkdir hf

export MKL_SERVICE_FORCE_INTEL=1

xtuner convert pth_to_hf ./internlm2_chat_1_8b_qlora_copy.py ./work_dirs/internlm2_chat_1_8b_qlora_copy/iter_24236.pth ./hf

![image](https://github.com/InternLM/xtuner/assets/63148804/52093973-ece5-41e8-953e-c311dba06b51)

并入基础模型：

将LoRA模型合并入基础模型：

rm -rf merged

xtuner convert merge /home/zhanghui/models/Shanghai_AI_Laboratory/internlm2-chat-1_8b ./hf ./merged --max-shard-size 2GB

![image](https://github.com/InternLM/xtuner/assets/63148804/5228aa71-b2b8-4e73-ae38-d2caf0b1b52e)

使用chat验证：
xtuner chat ./merged --temperature 0.8 --top-p 0.8 --repetition-penalty 1.002 --prompt-template internlm2_chat
![image](https://github.com/InternLM/xtuner/assets/63148804/40361eaa-3156-46ee-b83f-ba3ddb726f0b)

使用app.py验证应用：
这是正常的：
![image](https://github.com/InternLM/xtuner/assets/63148804/3ac345f5-040d-4336-a56b-080461dedecf)

这是循环的：
![image](https://github.com/InternLM/xtuner/assets/63148804/fad9819f-16a9-44a8-8d0a-6ad13a870965)

其中代码中已经做了如下处理：
<img width="710" alt="fa8b66a7c783aa9cb00d67901d499c3" src="https://github.com/InternLM/xtuner/assets/63148804/df01905d-e46d-4e82-82bf-8cf4eca62117">

![image](https://github.com/InternLM/xtuner/assets/63148804/0b0629cb-73ad-4373-8d27-1f219f2ed19b)

停止符也确认没问题：
![image](https://github.com/InternLM/xtuner/assets/63148804/c5c137b1-1fa0-43c0-ad54-07afc8f983cb)

确实是大概率事件。

**Comment by LZHgrla** [comment]
@zhanghui-china  Hi
可以尝试增大repetition_penalty

**Comment by zhanghui-china** [comment]
> @zhanghui-china Hi 可以尝试增大repetition_penalty

现在是1.002，增大到多少比较合适呢？

**Comment by LZHgrla** [comment]
@zhanghui-china  可以以0.001为步长逐步增大，配合着实际效果调整

**Comment by zhanghui-china** [comment]
> @zhanghui-china 可以以0.001为步长逐步增大，配合着实际效果调整

改为1.1
<img width="647" alt="d658d2e936af6f48c293a45ab07dde0" src="https://github.com/InternLM/xtuner/assets/63148804/005a2e28-6708-4163-bf50-59a19675f749">
有时候会出现eoa

改为 1.2
<img width="647" alt="d658d2e936af6f48c293a45ab07dde0" src="https://github.com/InternLM/xtuner/assets/63148804/433ed456-ab82-46a9-b49e-a1cf24cb834b">

<img width="590" alt="b71b8abb8b7a18a7b9c657488950c81" src="https://github.com/InternLM/xtuner/assets/63148804/b3d20c5e-ab54-4130-bce9-70c71d69ba88">
有时候会出现大量eoa重复。

**Comment by LZHgrla** [comment]
确认一下你所使用的webui的对话模版？

**Comment by zhanghui-china** [comment]
> 确认一下你所使用的webui的对话模版？

```
# Model
pretrained_model_name_or_path = '/home/zhanghui/models/Shanghai_AI_Laboratory/internlm2-chat-1_8b'
use_varlen_attn = False

# Data
data_path = './dataset/tran_dataset_0.json'
prompt_template = PROMPT_TEMPLATE.internlm2_chat
max_length = 2048
pack_to_max_length = True

# Scheduler & Optimizer
batch_size = 1  # per_device
accumulative_counts = 16
dataloader_num_workers = 0
max_epochs = 2
optim_type = AdamW
lr = 2e-4
betas = (0.9, 0.999)
weight_decay = 0
max_norm = 1  # grad clip
warmup_ratio = 0.03

```

**Comment by zhanghui-china** [comment]
[
[20240306_015337.log](https://github.com/InternLM/xtuner/files/14514050/20240306_015337.log)
](url)

老师，我上传了xtuner的日志。您看看能不能分析下。

**Comment by LZHgrla** [comment]
> > 确认一下你所使用的webui的对话模版？
> 
> ```
> # Model
> pretrained_model_name_or_path = '/home/zhanghui/models/Shanghai_AI_Laboratory/internlm2-chat-1_8b'
> use_varlen_attn = False
> 
> # Data
> data_path = './dataset/tran_dataset_0.json'
> prompt_template = PROMPT_TEMPLATE.internlm2_chat
> max_length = 2048
> pack_to_max_length = True
> 
> # Scheduler & Optimizer
> batch_size = 1  # per_device
> accumulative_counts = 16
> dataloader_num_workers = 0
> max_epochs = 2
> optim_type = AdamW
> lr = 2e-4
> betas = (0.9, 0.999)
> weight_decay = 0
> max_norm = 1  # grad clip
> warmup_ratio = 0.03
> ```

请确认你所使用的webui的对话配置，而非训练配置。

**Comment by zhanghui-china** [comment]
if base_model_type == 'internlm-chat-7b':
        generation_config = GenerationConfig(
            max_length=max_length)   #InternLM1
    elif base_model_type == 'internlm2-chat-1.8b':
        generation_config = GenerationConfig(
            max_length=max_length, top_p=0.8, temperature=0.8, repetition_penalty=1.17)   #InternLM2 1.8b need 惩罚参数
    else:
        generation_config = GenerationConfig(
            max_length=max_length, top_p=0.8, temperature=0.8, repetition_penalty=1.002)   #InternLM2 2 need 惩罚参数

**Comment by zhanghui-china** [comment]
> > > 确认一下你所使用的webui的对话模版？
> > 
> > 
> > ```
> > # Model
> > pretrained_model_name_or_path = '/home/zhanghui/models/Shanghai_AI_Laboratory/internlm2-chat-1_8b'
> > use_varlen_attn = False
> > 
> > # Data
> > data_path = './dataset/tran_dataset_0.json'
> > prompt_template = PROMPT_TEMPLATE.internlm2_chat
> > max_length = 2048
> > pack_to_max_length = True
> > 
> > # Scheduler & Optimizer
> > batch_size = 1  # per_device
> > accumulative_counts = 16
> > dataloader_num_workers = 0
> > max_epochs = 2
> > optim_type = AdamW
> > lr = 2e-4
> > betas = (0.9, 0.999)
> > weight_decay = 0
> > max_norm = 1  # grad clip
> > warmup_ratio = 0.03
> > ```
> 
> 请确认你所使用的webui的对话配置，而非训练配置。
具体参见 https://github.com/SmartFlowAI/TheGodOfCookery/blob/version-0.4-1.8b/app.py

**Comment by LZHgrla** [comment]
需要确保在你所使用的webui中，传给模型的文本，嵌套的是internlm2的模版，具体可参考

https://github.com/InternLM/InternLM/blob/main/chat/chat_format_zh-CN.md#基本结构

https://github.com/InternLM/xtuner/blob/770bac38bc905794eb38e53de4f54f98e30a77dc/xtuner/utils/templates.py#L24-L31

**Comment by zhanghui-china** [comment]
好的。我试下
